examples/custom-middleware/tenant-group-manipulation.md

```diff
@@ -0,0 +1,46 @@
+---
+description: Tenant group manipulation
+---
+
+# Tenant group manipulation
+
+We can change a default tenant and order of the tenants in the ROR menu by providing `defaultGroup` query string parameter before the login, and chenge the order of `availableGroups`
+thanks to the `enrichIdentitySessionMetadata` method.
+
+1. Declare `readonlyrest_kbn.custom_middleware_inject_file: 'path/to/custom_middleware_inject_file.js'` in the kibana.yml and declare `custom_middleware_inject_file.js`
+
+```js
+async function customMiddleware(req, res, next) {
+    const rorRequest = req.rorRequest;
+    const metadata =
+        req.rorRequest && req.rorRequest.getIdentitySession() && req.rorRequest.getIdentitySession().metadata;
+    const defaultGroup = 'infosec';
+
+    if (rorRequest.getPath() === '/login' && rorRequest.getMethod() === 'post') {
+        if (rorRequest.getBody().username === 'admin') {
+            rorRequest.setQuery('defaultGroup', defaultGroup);
+        }
+    }
+
+    if (metadata && rorRequest.getPath() === '/pkp/api/info') {
+        const availableGroups = metadata.availableGroups;
+        if (availableGroups.some(availableGroup => availableGroup === defaultGroup)) {
+            const index = availableGroups.indexOf(defaultGroup);
+            const groupAvailable = index !== -1;
+            if (groupAvailable) {
+                availableGroups.splice(index, 1);
+                availableGroups.unshift(defaultGroup);
+            }
+
+            rorRequest.enrichIdentitySessionMetadata({ availableGroups });
+        }
+    }
+
+    return next();
+}
+```
+
+In this example, before the login to the Kibana, when the username is equal 'admin', we add default tenant `rorRequest.setQuery('defaultGroup', defaultGroup);` which means,
+that it will be the first tenant opened after the login. After the login, we will change the order of tenants displayed in the ROR menu and our default tenant will be the first on the list.
+
+**⚠️IMPORTANT** Custom middleware must return `next()` function, to not block the request
```

Comment:
We should mention the limitations of this approach. This will not work when the ROR login form is not directly involved in the login process. This happens when users log in through SAML/OIDC or reverse proxy (`proxy_pass` rule).

Right?

Reply:
@Dzuming when this is merged, let's add a deep link from the changelog entry to this page (using the docs.readonlyrest.com URL rather than direct github of course)

Reply:
> We should mention the limitations of this approach. This will not work when the ROR login form is not directly involved in the login process. This happens when users log in through SAML/OIDC or reverse proxy (proxy_pass rule).

I'm not sure if i understand, what kind of problem can you see here? IMO No matter which, method to login the user will use, it will be always `POST /login` call which will be intercepted by the custom middleware. The only problem here, is that in case of SAML/OIDC to receive username, user needs to decrypt JWT token, but i can add it to the examples.

Reply:
Hmm what about the proxy pass? IIRC in ROR there's a middleware that catches X-Forwarded-User on the fly without requiring a POST /login?

Reply:
Good point, but in theory user could provide a default group by something like

```
  const metadata =
    req.rorRequest && req.rorRequest.getIdentitySession() && req.rorRequest.getIdentitySession().metadata;
  const defaultGroup = 'infosec';
 const X_FORWARDED_USER = 'x-forwarded-user';

  if (!metadata && rorRequest.getHeaders()[X_FORWARDED_USER]) {
    if (rorRequest.getHeaders()[X_FORWARDED_USER] === 'admin') {
      rorRequest.setQuery('defaultGroup', defaultGroup);
    }
```
Unfortunately, handleProxyAuth middleware is running before a custom middleware so it won't work, I will create a Jira for it and fix it. To sum up:

- I will provide a SAML/OIDC example
- I will add info that proxy auth not support it yet
- I will create a Jira to call custom middleware before proxy auth

Reply:
Updated ✔️